kafka/src/test/java/io/specmesh/kafka/ExporterFunctionalTest.java

```diff
@@ -0,0 +1,142 @@
+
+    private enum Domain {
+        /** The domain associated with the spec. */
+        SELF(API_SPEC.id()),
+        /** An unrelated domain. */
+        UNRELATED(".london.hammersmith.transport"),
+        /** A domain granted access to the protected topic. */
+        LIMITED(".some.other.domain.root");
+
+        final String domainId;
+
+        Domain(final String name) {
+            this.domainId = name;
+        }
+    }
+
+    private static final String ADMIN_USER = "admin";
+
+    @RegisterExtension
+    private static final KafkaEnvironment KAFKA_ENV =
+            DockerKafkaEnvironment.builder()
+                    .withoutSchemaRegistry()
+                    .withSaslAuthentication(
+                            ADMIN_USER,
+                            ADMIN_USER + "-secret",
+                            Domain.SELF.domainId,
+                            Domain.SELF.domainId + "-secret",
+                            Domain.UNRELATED.domainId,
+                            Domain.UNRELATED.domainId + "-secret",
+                            Domain.LIMITED.domainId,
+                            Domain.LIMITED.domainId + "-secret")
+                    .withKafkaAcls(aclsForOtherDomain(Domain.LIMITED))
+                    .withKafkaAcls(aclsForOtherDomain(Domain.UNRELATED))
+                    .build();
+
+    @BeforeAll
+    static void setUp() {
+        try (Admin adminClient = KAFKA_ENV.adminClient()) {
+            Provisioner.provisionTopics(API_SPEC, adminClient);
+            Provisioner.provisionAcls(API_SPEC, adminClient);
+        }
+    }
+
+    @Test
+    void shouldExportAPIFromCluster() throws Exporter.ExporterException {
+
+        try (Admin adminClient = KAFKA_ENV.adminClient()) {
+            final ApiSpec exported = new Exporter().export(aggregateId, adminClient);
+            assertThat(exported.channels(), is(notNullValue()));
+            assertThat(exported.channels().size(), is(3));
+            exported.channels()
+                    .keySet()
+                    .forEach(
+                            c ->
+                                    assertThat(
+                                            "Should natch aggregate level topic prefix",
+                                            c.matches(
+                                                    aggregateId.replace(':', '.')
+                                                            + ".(_public|_private|_protected).*")));
+        }
+    }
+
+    @Test
+    void shouldExportYAMLAPIFromCluster() throws Exporter.ExporterException {
+
+        try (Admin adminClient = KAFKA_ENV.adminClient()) {
+
+            final ApiSpec exported = new Exporter().export(aggregateId, adminClient);
+            final String specYaml = Exporter.exportYaml(exported);
+            assertThat(
+                    specYaml,
+                    is(containsString("id: \"urn:london:hammersmith:olympia:bigdatalondon\"")));
+        }
+    }
```

Comment:
This test does a great job of providing evidence that the `exportYaml` method belongs on another class.

This test: 
1. creates a Kafka cluster in Docker containers, 
2. applies a spec to it, 
3. exports the spec and then 
4. converts that spec to YAML. 

When all that is really necessary is to test this method is to create a spec and export it to YAML.  Move it to its own class.

Also, assert the spec is _exactly_ what you expect, not that it contains one tiny bit of the spec.  Otherwise it could be outputting complete garbage, but your test will pass...

Reply:
updated